clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsResponse.java

```diff
@@ -271,4 +273,48 @@ public static ApiVersion toApiVersion(ApiKeys apiKey) {
             .setMinVersion(apiKey.oldestVersion())
             .setMaxVersion(apiKey.latestVersion());
     }
+
+    public static ApiVersionsResponse apiVersionsResponse(
```

Comment:
Do we really need two constructors for this? It seems like initializing an `ApiVersionsResponse` without any finalized features set is something we'd only want to do in a test. And we can just have a convenience method in a test that has far fewer parameters than this one (plus won't get called accidentally somewhere else where we really do need finalized features set)

Reply:
There's one non-test use case [here](https://github.com/ahuang98/kafka/blob/0cb2a9230476654b60dd54ef0147bae0cf0404ea/core/src/main/scala/kafka/server/ApiVersionManager.scala#L84-L98) in `ApiVersionManager.apiVersionResponse` which is called by SocketServer, ControllerApis, KafkaApis. I guess even if it isn't safe to assume finalized features will be set, I can refactor the `ApiVersionManager.apiVersionResponse` to throw an exception instead of using the match statement?